starlarkstruct/struct.go

```diff
@@ -268,14 +230,31 @@ func structsEqual(x, y *Struct, depth int) (bool, error) {
 		return false, nil
 	}
 
-	for i, n := 0, x.len(); i < n; i++ {
-		if x.entries[i].name != y.entries[i].name {
-			return false, nil
-		} else if eq, err := starlark.EqualDepth(x.entries[i].value, y.entries[i].value, depth-1); err != nil {
-			return false, err
+	// Loop over every key returning the lowest error by key, if any.
```

Comment:
This loop must now visit every pair of elements even when the first pair differs. That's a regression. I'm not sure StringDict really makes sense as a representation for struct. The requirements are that:

1. struct is compact.
2. struct.f is fast.
3. str(struct) uses field-name order.
4. hash(struct) uses field-name order (or is order-independent).
5. struct==struct compares fields in name order.

The previous representation using a name-ordered k/v array was better on all requirements except 2, and even then the advantage is only material when N=16, which is an unusually large struct in a typical Starlark program.

I suspect the access patterns of Starlark fields in Attr are highly temporally localized: the most likely field name to be accessed is the previous one. If that is true (and it would be worth measuring), it might be possible to outperform both the existing implementation and the hash table by caching the resulting index of the binary search in a field in the struct (without a mutex!) and using it as a hint next time.

```
func FromStringDict()  {
   ...
   s := &Struct{
	constructor: constructor,
	hint: maxuint32,
   }
}

func (s *Struct) Attr(name string) (starlark.Value, error) {
        if hint := s.hint < n && s.entries[hint].name == name {
              return ... // yay the hint worked
        }
	// Binary search the entries.
	for i < j { ...binary search...}
	if i < n && s.entries[i].name == name {
                s.hint = i // save the hint
		return s.entries[i].value, nil
	}
        ...
}
```

What do you think?

Reply:
I agree. I think the original binary search implementation is better. Hint idea sounds good, I'll have a look at how to validate it. Please let me know if you'd like to keep the benchmarking code and I could modify this PR to just include that? Otherwise please close! Thanks for taking the time to review.

I did try a solution of based on the `starlark.hashtable` here: https://github.com/emcfarlane/starlark-go/pull/1 . It does look like it could beat both implementations but requires modify the original hashtable, and cloning it out of the package. Just an experiment.